packages/opds-web-client/src/hooks/useDownloadButton.ts

```diff
@@ -0,0 +1,95 @@
+import * as React from "react";
+import { MediaLink, FulfillmentLink, MediaType } from "./../interfaces";
+import { useActions } from "../components/context/ActionsContext";
+import download from "../components/download";
+import { generateFilename, typeMap } from "../utils/file";
+
+export function fixMimeType(
+  mimeType: MediaType | "vnd.adobe/adept+xml"
+): MediaType {
+  return mimeType === "vnd.adobe/adept+xml"
+    ? "application/vnd.adobe.adept+xml"
+    : mimeType;
+}
+
+function isIndirect(
+  link: MediaLink | FulfillmentLink
+): link is FulfillmentLink {
+  return !!(
+    (link as FulfillmentLink).indirectType &&
+    link.type === "application/atom+xml;type=entry;profile=opds-catalog"
+  );
+}
+
+const STREAMING_MEDIA_LINK_TYPE: MediaType =
+  "text/html;profile=http://librarysimplified.org/terms/profiles/streaming-media";
+
+type DownloadDetails = {
+  fulfill: () => Promise<void>;
+  isIndirect: boolean;
+  downloadLabel: string;
+  mimeType: MediaType;
+  fileExtension: string;
+};
+/**
+ * We use typescript function overloads to show that if you pass in a link
+ * you are guaranteed to get the download button details in an object. If
+ * you might pass in undefined, you might get null
+ */
+export default function useDownloadButton(
+  link: MediaLink | FulfillmentLink,
+  title: string
+): DownloadDetails;
+export default function useDownloadButton(
+  link: MediaLink | FulfillmentLink | undefined,
```

Comment:
Would this work if you did `link: undefined,` for this function declaration?

Reply:
For the declaration or for when you call the function? It allows you to call the function with a value that has type `MediaLink | FulfillmentLink | undefined`, but then it may return `null` (if in fact the value turns out to be `undefined`. However, if you pass it a value that is guaranteed to be `MediaLink | FulfillmentLink`, you are guaranteed to get back `DownloadDetails`. Does that make sense?